File: toothpick-compiler/src/main/java/toothpick/compiler/factory/FactoryGenerator.java
package toothpick.compiler.factory;

import com.squareup.javapoet.ClassName;
import com.squareup.javapoet.JavaFile;
import com.squareup.javapoet.MethodSpec;
import com.squareup.javapoet.ParameterizedTypeName;
import com.squareup.javapoet.TypeName;
import com.squareup.javapoet.TypeSpec;
import javax.lang.model.element.Modifier;
import javax.lang.model.type.TypeMirror;
import toothpick.Factory;
import toothpick.Injector;
import toothpick.compiler.CodeGenerator;
import toothpick.compiler.targets.ConstructorInjectionTarget;

/**
 * Generates a {@link Factory} for a given {@link ConstructorInjectionTarget}.
 * Typically a factory is created for a class a soon as it contains
 * an {@link javax.inject.Inject} annotated constructor.
 * TODO also generate a factory for a <em>non private & non abstract class</em>
 * whose instances are injected (i.e. the type of an injected field or of a parameter
 * of an injected constructor.
 */
public class FactoryGenerator implements CodeGenerator {

  private static final String FACTORY_SUFFIX = "$$Factory";

  private ConstructorInjectionTarget constructorInjectionTarget;

  public FactoryGenerator(ConstructorInjectionTarget constructorInjectionTarget) {
    this.constructorInjectionTarget = constructorInjectionTarget;
  }

  public String brewJava() {
    // Interface to implement
    ClassName className = ClassName.get(constructorInjectionTarget.builtClass);
    ParameterizedTypeName parameterizedTypeName = ParameterizedTypeName.get(ClassName.get(Factory.class), className);

    // Build class
    TypeSpec.Builder factoryTypeSpec = TypeSpec.classBuilder(className.simpleName() + FACTORY_SUFFIX)
        .addModifiers(Modifier.PUBLIC, Modifier.FINAL)
        .addSuperinterface(parameterizedTypeName);
    emitCreateInstance(factoryTypeSpec);
    emitHasSingleton(factoryTypeSpec);
    emitHasProducesSingleton(factoryTypeSpec);

    JavaFile javaFile = JavaFile.builder(className.packageName(), factoryTypeSpec.build()).build();
    return javaFile.toString();
  }

  @Override public String getFqcn() {
    return constructorInjectionTarget.builtClass.getQualifiedName().toString() + FACTORY_SUFFIX;
  }

  private void emitCreateInstance(TypeSpec.Builder builder) {
    ClassName className = ClassName.get(constructorInjectionTarget.builtClass);
    MethodSpec.Builder createInstanceBuilder = MethodSpec.methodBuilder("createInstance")
        .addAnnotation(Override.class)
        .addModifiers(Modifier.PUBLIC)
        .addParameter(ClassName.get(Injector.class), "injector")
        .returns(className);

    StringBuilder localVarStatement = new StringBuilder("");
    localVarStatement.append(className.simpleName()).append(" ");
    String varName = "" + Character.toLowerCase(className.simpleName().charAt(0));
    varName += className.simpleName().substring(1);
    localVarStatement.append(varName).append(" = ");
    localVarStatement.append("new ");
    localVarStatement.append(className.simpleName()).append("(");
    int counter = 1;
    String prefix = "";

    for (TypeMirror typeMirror : constructorInjectionTarget.parameters) {
      String paramName = "param" + counter++;
      TypeName paramType = TypeName.get(typeMirror);
      createInstanceBuilder.addStatement("$T $L = injector.getInstance($T.class)", paramType, paramName, paramType);
      localVarStatement.append(prefix);
      localVarStatement.append(paramName);
      prefix = ", ";
    }

    localVarStatement.append(")");
    createInstanceBuilder.addStatement(localVarStatement.toString());
    if (constructorInjectionTarget.needsMemberInjection) {
      StringBuilder injectStatement = new StringBuilder("injector.inject(");
      injectStatement.append(varName);
      injectStatement.append(")");
      createInstanceBuilder.addStatement(injectStatement.toString());
    }
    StringBuilder returnStatement = new StringBuilder("return ");
    returnStatement.append(varName);
    createInstanceBuilder.addStatement(returnStatement.toString());

    builder.addMethod(createInstanceBuilder.build());
  }

  private void emitHasSingleton(TypeSpec.Builder builder) {
    MethodSpec.Builder hasSingletonBuilder = MethodSpec.methodBuilder("hasSingletonAnnotation")
        .addAnnotation(Override.class)
        .addModifiers(Modifier.PUBLIC)
        .returns(TypeName.BOOLEAN)
        .addStatement("return $L", constructorInjectionTarget.hasSingletonAnnotation);
    builder.addMethod(hasSingletonBuilder.build());
  }

  private void emitHasProducesSingleton(TypeSpec.Builder builder) {
    MethodSpec.Builder hasProducesSingletonBuilder = MethodSpec.methodBuilder("hasProducesSingletonAnnotation")
        .addAnnotation(Override.class)
        .addModifiers(Modifier.PUBLIC)
        .returns(TypeName.BOOLEAN)
        .addStatement("return $L", constructorInjectionTarget.hasProducesSingletonAnnotation);
    builder.addMethod(hasProducesSingletonBuilder.build());
  }
}

File: toothpick-compiler-ksp-factory/src/main/java/toothpick/compiler/factory/generators/FactoryGenerator.kt
/*
 * Copyright 2022 Baptiste Candellier
 * Copyright 2019 Stephane Nicolas
 * Copyright 2019 Daniel Molinero Reguera
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package toothpick.compiler.factory.generators

import com.google.devtools.ksp.getVisibility
import com.google.devtools.ksp.symbol.KSClassDeclaration
import com.squareup.kotlinpoet.AnnotationSpec
import com.squareup.kotlinpoet.ClassName
import com.squareup.kotlinpoet.CodeBlock
import com.squareup.kotlinpoet.FileSpec
import com.squareup.kotlinpoet.FunSpec
import com.squareup.kotlinpoet.KModifier
import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy
import com.squareup.kotlinpoet.PropertySpec
import com.squareup.kotlinpoet.TypeSpec
import com.squareup.kotlinpoet.asClassName
import com.squareup.kotlinpoet.ksp.KotlinPoetKspPreview
import com.squareup.kotlinpoet.ksp.addOriginatingKSFile
import com.squareup.kotlinpoet.ksp.toClassName
import com.squareup.kotlinpoet.ksp.toKModifier
import toothpick.Factory
import toothpick.MemberInjector
import toothpick.Scope
import toothpick.compiler.common.generators.TPCodeGenerator
import toothpick.compiler.common.generators.factoryClassName
import toothpick.compiler.common.generators.memberInjectorClassName
import toothpick.compiler.common.generators.targets.getInvokeScopeGetMethodWithNameCodeBlock
import toothpick.compiler.common.generators.targets.getParamType
import toothpick.compiler.factory.targets.ConstructorInjectionTarget
import javax.inject.Singleton

/**
 * Generates a [Factory] for a given [ConstructorInjectionTarget].
 *
 * Typically, a factory is created for a class a soon as it contains an [javax.inject.Inject] annotated constructor.
 * See Optimistic creation of factories in TP wiki.
 */
@OptIn(KotlinPoetKspPreview::class)
internal class FactoryGenerator(
    private val constructorInjectionTarget: ConstructorInjectionTarget
) : TPCodeGenerator {

    private val sourceClass: KSClassDeclaration = constructorInjectionTarget.sourceClass

    val sourceClassName: ClassName = sourceClass.toClassName()
    val generatedClassName: ClassName = sourceClassName.factoryClassName

    override fun brewCode(): FileSpec {
        return FileSpec.get(
            packageName = sourceClassName.packageName,
            TypeSpec.classBuilder(generatedClassName)
                .addOriginatingKSFile(sourceClass.containingFile!!)
                .addModifiers(sourceClass.getVisibility().toKModifier() ?: KModifier.PUBLIC)
                .addSuperinterface(
                    Factory::class.asClassName().parameterizedBy(sourceClassName)
                )
                .addAnnotation(
                    AnnotationSpec.builder(Suppress::class)
                        .addMember("%S", "ClassName")
                        .addMember("%S", "RedundantVisibilityModifier")
                        .build()
                )
                .emitSuperMemberInjectorFieldIfNeeded()
                .emitCreateInstance()
                .emitGetTargetScope()
                .emitHasScopeAnnotation()
                .emitHasSingletonAnnotation()
                .emitHasReleasableAnnotation()
                .emitHasProvidesSingletonAnnotation()
                .emitHasProvidesReleasableAnnotation()
                .build()
        )
    }

    private fun TypeSpec.Builder.emitSuperMemberInjectorFieldIfNeeded() = apply {
        val superTypeThatNeedsInjection: ClassName =
            constructorInjectionTarget
                .superClassThatNeedsMemberInjection
                ?.toClassName()
                ?: return this

        PropertySpec
            .builder(
                "memberInjector",
                MemberInjector::class.asClassName().parameterizedBy(superTypeThatNeedsInjection),
                KModifier.PRIVATE
            )
            .initializer("%T()", superTypeThatNeedsInjection.memberInjectorClassName)
            .build()
            .also { addProperty(it) }
    }

    private fun TypeSpec.Builder.emitCreateInstance(): TypeSpec.Builder = apply {
        FunSpec.builder("createInstance")
            .addModifiers(KModifier.OVERRIDE)
            .addParameter("scope", Scope::class)
            .returns(sourceClassName)
            .apply {
                // change the scope to target scope so that all dependencies are created in the target scope
                // and the potential injection take place in the target scope too
                if (constructorInjectionTarget.parameters.isNotEmpty() ||
                    constructorInjectionTarget.superClassThatNeedsMemberInjection != null
                ) {
                    addAnnotation(
                        AnnotationSpec.builder(Suppress::class)
                            .addMember("%S", "NAME_SHADOWING")
                            .build()
                    )

                    // We only need it when the constructor contains parameters or dependencies
                    addStatement("val scope = getTargetScope(scope)")
                }
            }
            .addCode(
                CodeBlock.builder()
                    .apply {
                        constructorInjectionTarget.parameters.forEachIndexed { i, param ->
                            addStatement(
                                "val %N = scope.%L as %T",
                                "param${i + 1}",
                                param.getInvokeScopeGetMethodWithNameCodeBlock(),
                                param.getParamType()
                            )
                        }

                        addStatement(
                            "return %T(%L)",
                            sourceClassName,
                            List(constructorInjectionTarget.parameters.size) { i -> "param${i + 1}" }
                                .joinToString(", ")
                        )

                        if (constructorInjectionTarget.superClassThatNeedsMemberInjection != null) {
                            beginControlFlow(".apply")
                            addStatement("memberInjector.inject(this, scope)")
                            endControlFlow()
                        }
                    }
                    .build()
            )
            .build()
            .also { addFunction(it) }
    }

    private fun TypeSpec.Builder.emitGetTargetScope(): TypeSpec.Builder = apply {
        val scopeName = constructorInjectionTarget.scopeName
        FunSpec.builder("getTargetScope")
            .addModifiers(KModifier.OVERRIDE)
            .addParameter("scope", Scope::class)
            .returns(Scope::class)
            .addStatement(
                "return %L",
                when (scopeName?.asString()) {
                    null -> CodeBlock.of("scope")
                    // there is no scope name or the current @Scoped annotation.
                    Singleton::class.qualifiedName!! -> CodeBlock.of("scope.rootScope")
                    else -> CodeBlock.of(
                        "scope.getParentScope(%T::class.java)",
                        ClassName(
                            packageName = scopeName.getQualifier(),
                            scopeName.getShortName()
                        )
                    )
                }
            )
            .build()
            .also { addFunction(it) }
    }

    private fun TypeSpec.Builder.emitHasScopeAnnotation(): TypeSpec.Builder = apply {
        val scopeName = constructorInjectionTarget.scopeName
        val hasScopeAnnotation = scopeName != null
        FunSpec.builder("hasScopeAnnotation")
            .addModifiers(KModifier.OVERRIDE)
            .returns(Boolean::class)
            .addStatement("return %L", hasScopeAnnotation)
            .build()
            .also { addFunction(it) }
    }

    private fun TypeSpec.Builder.emitHasSingletonAnnotation(): TypeSpec.Builder = apply {
        FunSpec.builder("hasSingletonAnnotation")
            .addModifiers(KModifier.OVERRIDE)
            .returns(Boolean::class)
            .addStatement(
                "return %L",
                constructorInjectionTarget.hasSingletonAnnotation
            )
            .build()
            .also { addFunction(it) }
    }

    private fun TypeSpec.Builder.emitHasReleasableAnnotation(): TypeSpec.Builder = apply {
        FunSpec.builder("hasReleasableAnnotation")
            .addModifiers(KModifier.OVERRIDE)
            .returns(Boolean::class)
            .addStatement(
                "return %L",
                constructorInjectionTarget.hasReleasableAnnotation
            )
            .build()
            .also { addFunction(it) }
    }

    private fun TypeSpec.Builder.emitHasProvidesSingletonAnnotation(): TypeSpec.Builder = apply {
        FunSpec.builder("hasProvidesSingletonAnnotation")
            .addModifiers(KModifier.OVERRIDE)
            .returns(Boolean::class)
            .addStatement(
                "return %L",
                constructorInjectionTarget.hasProvidesSingletonAnnotation
            )
            .build()
            .also { addFunction(it) }
    }

    private fun TypeSpec.Builder.emitHasProvidesReleasableAnnotation(): TypeSpec.Builder = apply {
        FunSpec.builder("hasProvidesReleasableAnnotation")
            .addModifiers(KModifier.OVERRIDE)
            .returns(Boolean::class)
            .addStatement(
                "return %L",
                constructorInjectionTarget.hasProvidesReleasableAnnotation
            )
            .build()
            .also { addFunction(it) }
    }
}
